Mathlib/RingTheory/Valuation/ValuativeRel.lean

```diff
@@ -0,0 +1,582 @@
+
+end Valuation
+
+/-- A preorder on a ring is said to be "valuative" if it agrees with the
+valuative relation. -/
+class ValuativePreorder (R : Type*) [CommRing R] [ValuativeRel R] [Preorder R] where
+  dvd_iff_le (x y : R) : x ≤ᵥ y ↔ x ≤ y
+
+namespace ValuativeRel
+
+variable {R : Type*} [CommRing R] [ValuativeRel R]
+
+@[simp]
+lemma rel_refl (x : R) : x ≤ᵥ x := by
+  cases rel_total x x <;> assumption
+
+lemma rel_rfl {x : R} : x ≤ᵥ x :=
+  rel_refl x
+
+protected alias rel.refl := rel_refl
+
+protected alias rel.rfl := rel_rfl
+
+@[simp]
+theorem zero_rel (x : R) : 0 ≤ᵥ x := by
+  simpa using rel_mul_right x ((rel_total 0 1).resolve_right not_rel_one_zero)
+
+lemma rel_mul_left {x y : R} (z) : x ≤ᵥ y → (z * x) ≤ᵥ (z * y) := by
+  rw [mul_comm z x, mul_comm z y]
+  apply rel_mul_right
+
+instance : Trans (rel (R := R)) (rel (R := R)) (rel (R := R)) where
+  trans h1 h2 := rel_trans h1 h2
+
+lemma rel_mul {x x' y y' : R} : x ≤ᵥ y → x' ≤ᵥ y' → x * x' ≤ᵥ y * y' := by
+  intro h1 h2
+  calc x * x' ≤ᵥ x * y' := rel_mul_left _ h2
+    _ ≤ᵥ y * y' := rel_mul_right _ h1
+
+theorem rel_add_cases (x y : R) : x + y ≤ᵥ x ∨ x + y ≤ᵥ y :=
+  (rel_total y x).imp (fun h => rel_add .rfl h) (fun h => rel_add h .rfl)
+
+variable (R) in
+def unitSubmonoid : Submonoid R where
+  carrier := { x | ¬ x ≤ᵥ 0}
+  mul_mem' := by
+    intro x y hx hy
+    by_contra c
+    apply hy
+    simp only [Set.mem_setOf_eq, not_not] at c
+    rw [show (0 : R) = x * 0 by simp, mul_comm x y, mul_comm x 0] at c
+    exact rel_mul_cancel hx c
+  one_mem' := not_rel_one_zero
+
+@[simp]
+lemma unitSubmonoid_def (x : R) : x ∈ unitSubmonoid R ↔ ¬ x ≤ᵥ 0 := Iff.refl _
+
+@[simp]
+lemma right_cancel_unitSubmonoid (x y : R) (u : unitSubmonoid R) :
+    x * u ≤ᵥ y * u ↔ x ≤ᵥ y := by
+  refine ⟨fun h => rel_mul_cancel u.prop h, fun h => ?_⟩
+  exact rel_mul_right _ h
+
+@[simp]
+lemma left_cancel_unitSubmonoid (x y : R) (u : unitSubmonoid R) :
+    u * x ≤ᵥ u * y ↔ x ≤ᵥ y := by
+  rw [← right_cancel_unitSubmonoid x y u]
+  simp only [mul_comm _ x, mul_comm _ y]
+
+variable (R) in
+/-- The setoid used to construct `ValueMonoid R`. -/
+def valueSetoid : Setoid (R × unitSubmonoid R) where
+  r := fun (x,s) (y,t) => x * t ≤ᵥ y * s ∧ y * s ≤ᵥ x * t
+  iseqv := {
+    refl ru := ⟨rel_refl _, rel_refl _⟩
+    symm h := ⟨h.2, h.1⟩
+    trans := by
+      rintro ⟨r, u⟩ ⟨s, v⟩ ⟨t, w⟩ ⟨h1, h2⟩ ⟨h3, h4⟩
+      constructor
+      · have := rel_mul h1 (rel_refl ↑w)
+        rw [mul_right_comm s] at this
+        have := rel_trans this (rel_mul h3 (rel_refl _))
+        rw [mul_right_comm r, mul_right_comm t] at this
+        simpa using this
+      · have := rel_mul h4 (rel_refl ↑u)
+        rw [mul_right_comm s] at this
+        have := rel_trans this (rel_mul h2 (rel_refl _))
+        rw [mul_right_comm t, mul_right_comm r] at this
+        simpa using this
+  }
+
+variable (R) in
+/-- The "canonical" value monoid of a ring with a valuative relation. -/
+def ValueGroup := Quotient (valueSetoid R)
```

Comment:
I don't like the name `ValueGroup`. For a start, it's not a group. And second, mathematicians already use "value group" to mean something else, namely the units of this group with zero (which, not entirely coincidentally, is a group). I think that Lean's current ValuationRing.ValueGroup also stinks for the same reason. I think that going against the literature is a bad idea so I propose `ValueGroupWithZero` for this one. What do you think?